### PR TITLE
etcdserver: specify timeout caused by leader election

### PR DESCRIPTION
**Pull request by yichengq**
Before this PR, the timeout caused by leader election returns:

```
14:45:37 etcd2 | 2015-08-12 14:45:37.786349 E | etcdhttp: got unexpected
response error (etcdserver: request timed out)
```

After this PR:

```
15:16:16 etcd1 | 2015-08-12 15:16:16.663957 E | etcdhttp: got unexpected
response error (etcdserver: request may be dropped due to leader election)
15:16:16 etcd1 | 2015-08-12 15:16:16.679423 E | etcdhttp: got
unexpected response error (etcdserver: request may be dropped due to leader
election)
```

for #2905

### CODE REVIEW

File: etcdserver/server.go [hunk 1]
```diff
@@ -1014,3 +1015,19 @@ func (s *EtcdServer) updateClusterVersion(ver string) {
...
 	}
 }
+
+func (s *EtcdServer) parseProposeCtxErr(err error, start time.Time) error {
+	switch err {
+	case context.Canceled:
+		return ErrCanceled
+	case context.DeadlineExceeded:
+		curLeadElected := s.r.leadElectedTime()
+		prevLeadLost := curLeadElectied.Add(-2 * time.Duration(s.cfg.ElectionTicks) * time.Duration(s.cfg.TickMs) * time.Millisecond)
```
Comment (xiang90) [inline etcdserver/server.go]: this lost is not accurate. just for your information.

**Comment by yichengq** [comment]
PTAL

File: etcdserver/raft.go [hunk 2]
```diff
@@ -86,6 +87,9 @@ type raftNode struct {
 	term  uint64
 	lead  uint64
 
+	mu          sync.Mutex
+	leadElected time.Time
```
Comment (xiang90) [inline etcdserver/raft.go]: leadElectedTime

**Comment by xiang90** [comment]
LGTM

File: etcdserver/raft.go [hunk 3]
```diff
@@ -86,6 +87,9 @@ type raftNode struct {
 	term  uint64
 	lead  uint64
 
+	mu              sync.Mutex
+	leadElectedTime time.Time
```
Comment (xiang90) [inline etcdserver/raft.go]: // last lead elected time
lt time.Time

File: etcdserver/raft.go [hunk 4]
```diff
@@ -187,6 +196,12 @@ func (r *raftNode) apply() chan apply {
 	return r.applyc
 }
 
+func (r *raftNode) curLeadElectedTime() time.Time {
```
Comment (xiang90) [inline etcdserver/raft.go]: leadElectedTime

**Pull request merged**